src/Types.jl

```diff
@@ -0,0 +1,258 @@
+export TheoryPetriNet, PetriNet, AbstractPetriNet, ns, nt, ni, no,
+  add_species!, add_transition!, add_transitions!,
+  add_input!, add_inputs!, add_output!, add_outputs!, inputs, outputs,
+  TransitionMatrices, vectorfield,
+  TheoryLabelledPetriNet, LabelledPetriNet, AbstractLabelledPetriNet, sname, tname,
+  TheoryReactionNet, ReactionNet, AbstractReactionNet, concentration, concentrations, rate, rates,
+  TheoryLabelledReactionNet, LabelledReactionNet, AbstractLabelledReactionNet
+
+using Catlab
+using Catlab.CategoricalAlgebra.CSets
+using Catlab.Present
+using Catlab.Theories
+using Petri
+using LinearAlgebra: mul!
+import Petri: Model
+
+vectorify(n) = begin
+  if !(typeof(n) <: Union{Vector,Tuple}) || (typeof(n) <: Tuple && length(n) == 1)
+    [n]
+  else
+    n
+  end
+end
+
+state_dict(n) = Dict(s=>i for (i, s) in enumerate(n))
+
+# Petri Nets
+############
+
+@present TheoryPetriNet(FreeSchema) begin
+  T::Ob
+  S::Ob
+  I::Ob
+  O::Ob
+
+  it::Hom(I,T)
+  is::Hom(I,S)
+  ot::Hom(O,T)
+  os::Hom(O,S)
+end
+
+const AbstractPetriNet = AbstractACSetType(TheoryPetriNet)
+const PetriNet = CSetType(TheoryPetriNet,index=[:it,:is,:ot,:os])
+
+# PetriNet([:S, :I, :R], :infection=>((1, 2), 3))
+
+PetriNet(n,ts...) = begin
+  p = PetriNet()
+  add_species!(p, n)
+  add_transitions!(p, length(ts))
+  for (i,(ins,outs)) in enumerate(ts)
+    ins = vectorify(ins)
+    outs = vectorify(outs)
+    add_inputs!(p, length(ins), repeat([i], length(ins)), collect(ins))
+    add_outputs!(p, length(outs), repeat([i], length(outs)), collect(outs))
+  end
+  p
+end
+
+ns(p::AbstractPetriNet) = nparts(p,:S)
+nt(p::AbstractPetriNet) = nparts(p,:T)
+ni(p::AbstractPetriNet) = nparts(p,:I)
+no(p::AbstractPetriNet) = nparts(p,:O)
+
+add_species!(p::AbstractPetriNet;kw...) = add_part!(p,:S;kw...)
+add_species!(p::AbstractPetriNet,n;kw...) = add_parts!(p,:S,n;kw...)
+
+add_transition!(p::AbstractPetriNet;kw...) = add_part!(p,:T;kw...)
+add_transitions!(p::AbstractPetriNet,n;kw...) = add_parts!(p,:T,n;kw...)
+
+add_input!(p::AbstractPetriNet,t,s;kw...) = add_part!(p,:I;it=t,is=s,kw...)
+add_inputs!(p::AbstractPetriNet,n,t,s;kw...) = add_parts!(p,:I,n;it=t,is=s,kw...)
+
+add_output!(p::AbstractPetriNet,t,s;kw...) = add_part!(p,:O;ot=t,os=s,kw...)
+add_outputs!(p::AbstractPetriNet,n,t,s;kw...) = add_parts!(p,:O,n;ot=t,os=s,kw...)
+
+# Note: although indexing makes this pretty fast, it is often faster to bulk-convert
```

Comment:
Out of curiosity, did you run benchmarks? At some point we should benchmark C-sets to make sure they perform as well in reality as they should in theory.

Reply:
I haven't yet, with it fully implemented I am interested to see how it performs compared to the Petri.jl implementation.